homeassistant/components/backup/util.py

```diff
@@ -57,3 +62,45 @@ def read_backup(backup_path: Path) -> AgentBackup:
             protected=cast(bool, data.get("protected", False)),
             size=backup_path.stat().st_size,
         )
+
+
+async def receieve_file(
+    hass: HomeAssistant, contents: aiohttp.BodyPartReader, path: Path
+) -> None:
+    """Receive a file from a stream and write it to a file."""
+    queue: SimpleQueue[tuple[bytes, asyncio.Future[None] | None] | None] = SimpleQueue()
+
+    def _sync_queue_consumer() -> None:
+        with path.open("wb") as file_handle:
+            while True:
+                if (_chunk_future := queue.get()) is None:
+                    break
+                _chunk, _future = _chunk_future
+                if _future is not None:
+                    hass.loop.call_soon_threadsafe(_future.set_result, None)
+                file_handle.write(_chunk)
+
+    fut: asyncio.Future[None] | None = None
+    try:
+        fut = hass.async_add_executor_job(_sync_queue_consumer)
+        megabytes_sending = 0
+        while chunk := await contents.read_chunk(BUF_SIZE):
+            megabytes_sending += 1
+            if megabytes_sending % 5 != 0:
+                queue.put_nowait((chunk, None))
+                continue
+
+            chunk_future = hass.loop.create_future()
```

Comment:
Side note: Why is it important to wait until the chunk is written to disk? Is it to not consume the stream in vain and waste resources?

Reply:
I agree this is strange. It seems like a more obvious solution would be a sempahore initiated at 5 to ensure we don't fill the RAM if http reading is faster than file writing.